For Mentor Review | Output and error handling | SaadiaELF

SaadiaELF opened the pull request:
# Learners, PR Template

## Self checklist

- [x] I have titled my PR with `For Mentor Review | {Project Name} | {My Name}`
- [ ] I have included a list of changes
- [x] I have tested my changes
- [ ] I have explained my changes
- [x] I have asked someone to review my changes

## Description

### Change list

- Created a main.go file in a client folder to hold Client code
- Wrote main function handling response cases depending on the statusCode

## Rationale, briefly

- What is the goal of these changes?
- How did these changes address this goal or goals?
- Why are they a good idea?

## @mentions, #links

- Who needs to know about this? @mention them
- Is this related to another issue, PR, or repo? #link them

SaadiaELF committed:
connecting client to the server

SaadiaELF committed:
Merge branch 'CodeYourFuture:main' into project01-output-and-error-handling

SaadiaELF committed:
Add 200 resp

SaadiaELF committed:
Merge branch 'project01-output-and-error-handling' of github.com:SaadiaELF/immersive-go-course into project01-output-and-error-handling

SaadiaELF committed:
refactoring

SaadiaELF committed:
refactoring

SaadiaELF committed:
developed cases

SaadiaELF committed:
Added "while" scenario

SaadiaELF committed:
Refactoring

SaadiaELF committed:
Adding comments

berkeli commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,73 @@
+
+func main() {
+	// Connect to the server and getting a response
+	resp, err := http.Get("http://localhost:8080")
+	// Show error message if connection is not established
+	if err != nil {
+		handleError()
+	}
```
Should we inform the user of the actual error that was returned?

berkeli commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,73 @@
+	body, err := io.ReadAll(resp.Body)
+
+	// Show error message if we cannot read the response body
+	if err != nil {
+		handleError()
+	}
```
Same question as above

berkeli commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,73 @@
+		handleError()
+	}
+	// Convert response body from binary to string
+	sb := string(body)
```
This logic is fine, but it only gets used if the status is 200. So in every other scenario (500, 429) it's an unnecessary operation.

berkeli commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,73 @@
+	case 500:
+		fmt.Fprint(os.Stderr, sb+"\n")
+	default:
+		handleError()
```
This will produce a very ambigious error that will be hard to debug

berkeli commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,73 @@
+	case 429:
+		handleRateLimited(resp.Header.Get("Retry-After"))
+	case 500:
+		fmt.Fprint(os.Stderr, sb+"\n")
```
If response is 500, body might be empty (most of the time!). So this won't produce a very useful error log

berkeli commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,73 @@
+// Handle response and retry depending on the Retry-After header
+func handleRateLimited(retryTime string) {
+	retrySeconds := 0
+	var err = error(nil)
```
a better pattern is to declare as `var err error`

berkeli commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,73 @@
+func handleRateLimited(retryTime string) {
+	retrySeconds := 0
+	var err = error(nil)
+	retryTimeDate, err := time.Parse(time.RFC1123, retryTime)
```
should we check for errors here?

berkeli replied:
ah I see you check for errors below, but a better pattern would be to check for errors right after they are given a value.

berkeli commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,73 @@
+	} else {
+		retrySeconds, err = strconv.Atoi(retryTime)
+		if err != nil {
+			handleError()
```
same comment here on amigious errors, it will not be very clear to the user which part of the application produced the error

berkeli commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,73 @@
+	if retrySeconds > 1 && retrySeconds <= 5 {
+		fmt.Printf("We will retry to get you the weather. Please wait %d seconds\n", retrySeconds)
+		time.Sleep(time.Duration(retrySeconds) * time.Second)
+		main()
```
It's not a good pattern to call `main()` like this, it should only be called when you app is executed. Can you think of a way to rework this?

You might want to consider extracting the logic to make the request into a separate function and retry based on the return of that function inside a for loop

berkeli reviewed (commented):
Good work so far, I left a few comments on things to consider. Some high level points:
- consider what errors are produced by your app and how useful they might be to someone who doesn't have access to the source code
- we should not call `main()` function recursively as you've done. Some of the reasons: 
   - This makes unit testing very difficult, because it will require you to test the logic for the whole app.
   - main() function will usually read args, env variables and do basic initializations. Which can also interfere with tests and other parts of the application

SaadiaELF committed:
Refactoring error messages

SaadiaELF committed:
Fixed main recursif call

SaadiaELF committed:
Adding retries

SaadiaELF committed:
Add comments

SaadiaELF commented:
> Good work so far, I left a few comments on things to consider. Some high level points:
> 
> * consider what errors are produced by your app and how useful they might be to someone who doesn't have access to the source code
> * we should not call `main()` function recursively as you've done. Some of the reasons:
>   
>   * This makes unit testing very difficult, because it will require you to test the logic for the whole app.
>   * main() function will usually read args, env variables and do basic initializations. Which can also interfere with tests and other parts of the application

Thanks for taking time and reviewing my code , I made those changes : 

- made error messages more descriptive 
- Deleted the recursive call of `main()` and replaced it with a `for` loop that run depending on the number of retries

SaadiaELF committed:
refactoring

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,92 @@
+		body, err := io.ReadAll(resp.Body)
+		// Show error message if we cannot read the response body
+		if err != nil {
+			fmt.Print("Sorry we cannot get the weather!\n")
```
I would typically expect this kind of failure message to go to stderr rather than stdout - see https://github.com/CodeYourFuture/immersive-go-course/tree/main/projects/output-and-error-handling#when-to-write-to-standard-outerror

Also, I'd consider using `fmt.Println` rather than using `fmt.Print` and manually adding the `\n` at the end of the string you're printing - personally I almost never have a use-case for `fmt.Print`, I always end up either using `fmt.Printf` or `fmt.Println` (but sometimes I do still :))

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,92 @@
+	if err != nil {
+		fmt.Print("Sorry we cannot get the weather!\n")
+		fmt.Fprintf(os.Stderr, "Failed to make http request: %v\n", err)
+		os.Exit(1)
```
It looks like you're doing this error handling the same in your main function and this one - you probably only need one copy of it :) Also, have a look at the "Only call os.Exit from your main function." section of https://github.com/CodeYourFuture/immersive-go-course/tree/main/projects/output-and-error-handling#handling-errors-in-your-code

(You're doing this in a few other places too :))

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,92 @@
+		switch resp.StatusCode {
+		case 200:
+			// Set the retries number to 0 so we can run the program many time when the response is 200
+			retries = 0
```
You unconditionally `os.Exit` in this branch, so you probably don't need to reset `retries` here, but it's a good thing to think about in other contexts - resetting retry counts on successful responses often makes sense (and it's unusual, particularly in a server, for us to `os.Exit`, though it's the correct thing to do here)!

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,92 @@
+
+			// Convert response body from binary to string
+			sb := string(body)
+			fmt.Fprint(os.Stdout, sb+"\n")
```
Like above, a `fmt.Println(os.Stdout, sb)` avoids you needing to add the newline yourself

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,92 @@
+			handleRateLimited(resp.Header.Get("Retry-After"), retries)
+		case 500:
+			fmt.Print("Sorry we cannot get the weather!\n")
+			fmt.Fprint(os.Stderr, "Internal Server Error\n")
```
You _could_ choose to append the response body to this message you're printing, or you could choose not to. There are advantages to both ways - what do you think the trade-offs are here about doing so or not?

SaadiaELF replied:
I think including response body may give more context about the error and help debugging server-side but giving a user friendly message instead of the response body  - which contain technical details - will lead to a better user experience

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,92 @@
+			os.Exit(1)
+		default:
+			fmt.Print("Sorry we cannot get the weather!\n")
+			fmt.Fprint(os.Stderr, "Unexpected Error\n")
```
I would probably at least include the status code you encountered here to help in debugging (and maybe more information or maybe not :))

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,92 @@
+		retrySeconds = 5
+	} else {
+		retrySeconds, err = strconv.Atoi(retryTime)
+		if err != nil {
```
You probably want to propagate an error here rather than handling it yourself

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,92 @@
+	}
+	if retrySeconds > 1 && retrySeconds <= 5 {
+		fmt.Printf("We will retry to get you the weather. Please wait %d seconds\n", retrySeconds)
+		time.Sleep(time.Duration(retrySeconds) * time.Second)
```
This sleep is going to make it quite slow to run any tests you may write about parsing this header - can you think of a way you could refactor the code so you can _quickly_ run tests for your parsing logic?

SaadiaELF replied:
I'm thinking of using milliseconds instead of seconds , but I may need to store the limits in variables too . is it the right approach ?

illicitonion replied:
I would think about separating the "work out how long we need to sleep" function from the "actually sleep" part - if you introduced a separate function for the `string` -> `time.Duration` calculation which you call in this function, you can test that without actually needing to sleep.

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,92 @@
+			sb := string(body)
+			fmt.Fprint(os.Stdout, sb+"\n")
+			os.Exit(0)
+		case 429:
+			handleRateLimited(resp.Header.Get("Retry-After"), retries)
```
Can you add this logging and check that `retries` is working like you expect? If not, can you do some debugging as to why?

```suggestion
		case 429:
			handleRateLimited(resp.Header.Get("Retry-After"), retries)
			fmt.Fprintf(os.Stderr, "After rate-limited response, retries is: %d\n", retries)
```

SaadiaELF replied:
Yes , it wasn't working as expected ( the `retries` variable never get incremented) because the incrementation was inside a local scope. But we want to see the effect in the `main` function. Fixed by moving `retries++` to `main` just before closing the `for loop`

illicitonion reviewed (commented):
This is looking good, and is a big improvement on the first version you put together - congrats!

I've left comments on a few things to think about - I'd encourage you to think about and address/discuss them, and then try to write some tests for this code - writing tests will help you think about advantages and disadvantages of different ways of structuring your code, and some of the things like why we don't want to `os.Exit(1)` in helper functions :)

SaadiaELF committed:
Refactoring repetitive error msg

SaadiaELF committed:
Deleting os.exit wrong calls

SaadiaELF committed:
Include status code

SaadiaELF committed:
Adding error propagation

SaadiaELF committed:
Fixed retries incrementation

SaadiaELF committed:
Add http client

SaadiaELF committed:
Added first test case

SaadiaELF committed:
Add 2 success cases

SaadiaELF committed:
Refactoring for loop

SaadiaELF committed:
Refactoring case 200

SaadiaELF committed:
Refactoring switch case

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,100 @@
+			os.Exit(1)
+		}
+		// Close the response body after been fully read
+		defer resp.Body.Close()
```
FYI `defer`s get called at the end of the _function_, whereas I suspect you really want this to happen at the end of the _loop iteration_. In this instance it doesn't really matter (the program will be done soon enough regardless), but it's good to know e.g. if you're writing a server that loops for a long time you'll end up leaking memory if you `defer` inside a loop.

To fix this, you could pull out a function which does "the contents of the for loop" :)

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,100 @@
+	// Create an HTTP client
+	client := &http.Client{}
+	// This loop will run only if the number of reties doesn't exceed 3
+	for retries := 0; retries < maxRetries; retries++ {
```
What will happen if the server _always_ succeeds (i.e. every response it gives is a 200)?

SaadiaELF replied:
Since I didn't implement `os.Exit(0)` yet , the loop doesn't stop on success. I'm trying to workout how to implement it since I should only call it from `main` :)

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,100 @@
+	retryTimeDate, err := time.Parse(time.RFC1123, retryTime)
+
+	if err == nil {
+		retrySeconds = int(time.Until(retryTimeDate).Seconds())
```
What will your code do if the timestamp given is the same as the current time? Or a second or two ago? What do you think we _should_ do in that case?

This can happen because of latency in getting a response, but also note that different computers' clocks may not be in sync!

SaadiaELF committed:
Refactoring handleRateLimited func

illicitonion commented on projects/output-and-error-handling/client/main_test.go:
```diff
@@ -0,0 +1,34 @@
+	"testing"
+)
+
+func TestGetWeather_Success(t *testing.T) {
```
This code doesn't really test your code very much - it's more of a re-implementation of your code (which calls your existing `getWeather` function).

To test your code you probably want:
1. To test your `handleRateLimited` function (or an extracted string -> time.Duration function).
2. To pull out another function or two that have more testable interfaces, e.g. if you extract a function for the body of your for loop that returns a `string, error`, you could test that the string it returns is the correct one in the success case, and that the error is what you expect in an error case.

SaadiaELF committed:
Refactoring convertTime func

illicitonion reviewed (commented):
This is looking much cleaner, congrats on the progress!

(It also looks like you've started pulling out the `convertTime` function I was suggesting as I was reviewing - nice!)

SaadiaELF committed:
Refactoring handleWeatherRequest func

SaadiaELF committed:
Clean up comments

SaadiaELF committed:
Refactoring

SaadiaELF committed:
Add InternalServerError test

SaadiaELF committed:
Add handleRateLimited tests

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,110 @@
+	client := &http.Client{}
+
+	for retries := 0; retries <= maxRetries; retries++ {
+		response, err := handleWeatherRequest(url, client, currentTime, retries)
```
"handle" to me implies "given something's been done, handle it" - I'd probably call this `makeWeatherRequest` because your function actually _makes_ the request, it doesn't just handle something

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,110 @@
+)
+
+func main() {
+	currentTime := time.Now()
```
This is maybe _slightly_ early to be getting the current time - ideally you probably want to get the time at around the same time as you get a response.

Imagine you get the current time as 10:00:00, then are told to sleep until 10:00:04, then retry and are told to sleep until 10:00:07 - your current code would sleep for a total of 11 seconds, where you actually only want to sleep for 7 seconds.

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,110 @@
+func handleWeatherRequest(url string, client *http.Client, currentTime time.Time, retries int) (string, error) {
+	resp, err := getWeather(url, client)
+	if err != nil {
+		return "", fmt.Errorf("failed to get weather: %v", err)
```
Prefer to use `%w` when wrapping errors:

```suggestion
		return "", fmt.Errorf("failed to get weather: %w", err)
```

See https://stackoverflow.com/questions/61283248/format-errors-in-go-s-v-or-w for details

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,110 @@
+	for retries := 0; retries <= maxRetries; retries++ {
+		response, err := handleWeatherRequest(url, client, currentTime, retries)
+		if err != nil {
+			handleError(err)
```
Now that you only call this in one place, I'd probably inline it here.

SaadiaELF replied:
I did following you comment about `defer` , did I misunderstood it ?

> FYI defers get called at the end of the function, whereas I suspect you really want this to happen at the end of the loop iteration. In this instance it doesn't really matter (the program will be done soon enough regardless), but it's good to know e.g. if you're writing a server that loops for a long time you'll end up leaking memory if you defer inside a loop.
> 
> To fix this, you could pull out a function which does "the contents of the for loop" :)

illicitonion replied:
Yeah, for the rest of the behaviour I think this looks great, but `handleError` doesn't have any `defer`s in it, so it doesn't need to be pulled out into a separate function - it's fine if its code just sits inside the loop.

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,110 @@
+		time.Sleep(time.Duration(retrySeconds) * time.Second)
+		return nil
+	} else {
+		return fmt.Errorf("internal Error : Failed to retry")
```
There are a few quite different causes that may lead us to this `else`, which may have different responses - can you split these into slightly more specific error messages?

illicitonion commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,110 @@
+	}
+}
+
+func convertTime(retryTime string, currentTime time.Time) (int, error) {
```
There are a few things I'm not sure about implementation-detail-wise in this function - can you write some tests for it to make sure it works for all of:
1. Two different timestamps
2. Two different integer numbers of seconds

?

illicitonion commented on projects/output-and-error-handling/client/main_test.go:
```diff
@@ -0,0 +1,84 @@
+}
+
+func TestHandleRateLimited(t *testing.T) {
+	testCases := []int{3, 6, -1}
```
I'd probably pull this out as a table-driven test so that the expectations are alongside the input values:

```go
type testCase struct {
    headerValue string
    wantValue time.Duration
    wantError error
}

testCases := []testCase{
    testCase{
        headerValue: "3",
        wantValue: 3 * time.Second,
        wantError: nil
    }
}
...
```

This avoids having so much logic in your test, and having to change the logic in the test if the input test-cases change.

Also, I'd probably mostly test this behaviour via a unit test of `convertTime`, and maybe just have one test that tests this via an actual HTTP request, rather than exclusively testing it via HTTP requests.

SaadiaELF replied:
> Also, I'd probably mostly test this behaviour via a unit test of `convertTime`, and maybe just have one test that tests this via an actual HTTP request, rather than exclusively testing it via HTTP requests.

If I have to test via `convertTime` function only how can I test the other error messages ? e.g the error on the `HandleRateLimited` ?

illicitonion replied:
Either with a separate test, or deciding not to bother :)

We don't need to just write one or two tests to try to cover all behaviours of our program, we can write lots of separate tests for different aspects of functionality.

illicitonion reviewed (commented):
Getting better each iteration!

SaadiaELF committed:
Fixed current time issue

SaadiaELF committed:
Improve handleRatelimit func

SaadiaELF committed:
Add timeProvider

SaadiaELF committed:
Add converTime test

SaadiaELF committed:
Add more test case to TestHandleRateLimited

SaadiaELF committed:
Tidy up

berkeli commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,124 @@
+	}
+}
+
+func makeWeatherRequest(url string, client *http.Client, retries int) (string, error) {
```
Small nit: usually, we like to order parameters in terms of significance (regardless of language). Which one do you think is the most significant?

berkeli replied:
same for `getWeather`  and `convertTime` functions

berkeli commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,124 @@
+	}
+}
+
+type TimeProvider interface {
+	Now() time.Time
+	Until(time.Time) time.Duration
+}
```
good use of interfaces!

berkeli commented on projects/output-and-error-handling/client/main_test.go:
```diff
@@ -0,0 +1,167 @@
+	}
+}
+
+func (m MockTimeProvider) Now() time.Time {
+	return time.Date(2024, time.March, 22, 12, 0, 0, 0, time.UTC)
+}
+
+func (m MockTimeProvider) Until(t time.Time) time.Duration {
+	return t.Sub(m.Now())
+}
```
I would move these methods so they are next to the struct. There are also libraries to create mocks:
* https://github.com/uber-go/mock
* https://github.com/vektra/mockery

They can be a little hard to get your head around at first, so let me know if you'd like to have a session or some examples

This is only for your knowledge, don't worry about them too much just yet

SaadiaELF replied:
I don't think I can use any library yet. I'm still trying to have a better understanding of Go basic concepts.

berkeli commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -0,0 +1,124 @@
+	return time.Until(t)
+}
+
+func convertTime(retryTime string, tp TimeProvider) (int, error) {
```
I think you can re-work this a little so it returns `time.Duration`, it will save you back and forth conversions

SaadiaELF replied:
I tried to return `time.Duration` but did a lot of changes , can you please have a look and see if I'm doing it in wrong way?.

berkeli replied:
No it seems correct to me! good work 👍

SaadiaELF committed:
Refactoring

berkeli commented on projects/output-and-error-handling/client/main.go:
```diff
@@ -105,19 +106,20 @@ func (r RealTimeProvider) Until(t time.Time) time.Duration {
+		retryDuration := time.Duration(secondsInt) * time.Second
+		return retryDuration, nil
 	} else if _, err := time.Parse(http.TimeFormat, retryTime); err == nil {
 		httpTime, err := time.Parse(http.TimeFormat, retryTime)
```
Here you can use the output from the first Parse, instead of parsing again
```go
else if val, err := time.Parse(http.TimeFormat, retryTime); err == nil {
		httpTime = val
```
this way, you also don't need to check the error repeatedly
